File: 21-2.py
#!/usr/bin/env python

from functools import reduce

specs = [spec.strip() for spec in open('21.input').readlines()]

foods = []
possible_allergens = {}

for spec in specs:
  ingredients, allergens = spec.split('(')
  ingredients = ingredients.split()
  allergens = [a.rstrip(',)') for a in allergens.split() if a != 'contains' ]
  for a in allergens:
    possible_allergens[a] = []
  foods.append([ingredients, allergens])

for allergen in possible_allergens.keys():
  foods_containing_allergen = [f[0] for f in foods if allergen in f[1]]
  possible_allergens[allergen] = reduce(lambda a, b: list(set(a) & set(b)), foods_containing_allergen)

actual_allergens = {}

while len(possible_allergens) > 0:
  definitive = [[allergen, possible_ingredients] for allergen, possible_ingredients in possible_allergens.items() if len(possible_ingredients) == 1]
  for allergen, ingredient in definitive:
    actual_allergens[allergen] = ingredient[0]
    for a, possible_ingredients in possible_allergens.items():
      if ingredient[0] in possible_ingredients:
        ingredients = possible_allergens[a]
        possible_allergens[a] = [i for i in ingredients if i != ingredient[0]]
    del possible_allergens[allergen]

result = ""
for allergen in sorted(actual_allergens.keys()):
  result += actual_allergens[allergen] + ","
print(result.rstrip(","))

File: 21-2.rb
#!/usr/bin/env ruby

specs = File.read('21.input').lines.map &:strip

foods = []
possible_allergens = {}

specs.each do |spec|
  ingredients, allergens = spec.split '('
  ingredients = ingredients.split
  allergens = allergens.split.map { |a| a.delete ',)' }.reject { |a| a == 'contains' }
  allergens.each do |a|
    possible_allergens[a] = []
  end
  foods.push [ingredients, allergens]
end

possible_allergens.each_key do |allergen|
  foods_containing_allergen = foods.select { |f| f[1].include? allergen }.map &:first
  possible_allergens[allergen] = foods_containing_allergen.reduce { |memo, f| memo.intersection f }
end

actual_allergens = {}

until possible_allergens.empty?
  definitive = possible_allergens.select do |allergen, possible_ingredients|
    possible_ingredients.size == 1
  end
  definitive.each do |allergen, ingredient|
    actual_allergens[allergen] = ingredient[0]
    possible_allergens.each do |a, possible_ingredients|
      if possible_ingredients.include? ingredient[0]
        possible_allergens[a] = possible_allergens[a].reject { |i| i == ingredient[0] }
      end
    end
    possible_allergens.delete allergen
  end
end

print actual_allergens.to_a.sort { |a, b| a[0] <=> b[0] }.map { |x| x[1] }.join ','
print "\n"
